Added validation on note creation.

nadiia-kobylinska opened the pull request:
- [ ] Note title cannot be empty
- [ ] Limit the content field to 1000 characters
- [ ] Show how many characters the text entry limit is exceeded
- [ ] Do not allow to save the note if it is not valid

Improved the code a bit by replacing useState with useReducer.

nadiia-emit-digital committed:
update validation

sologubd reviewed (approved):
Looks good to me.
The only one thing. I suggest using a code formatter in order to have the same spaces and indents everywhere.

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,99 +2,106 @@ import {Box, Button, Stack, Typography} from "@mui/material";
-        if (props.id && props.id !== editMode) {
-            setEditMode(props.id);
+        }else if (props.id !== state.editMode) {
+            dispatch({type: 'change', payload: { editMode: props.id}})
```
Предыдущий подход с условными операторами мне нравился больше. Если должно выполняться какое то одно условие то можно сделать ретурн в первом ифе.

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,99 +2,106 @@ import {Box, Button, Stack, Typography} from "@mui/material";
 
     useEffect(()=>{
-        if (props.id && editMode && props.id === editMode) {
+        if (props.id && state.editMode && props.id === state.editMode) {
```
Удобнее будет читать если сделать условие которое остановит выполнение по невалидным критериям. А саму логику писать уже после

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,99 +2,106 @@ import {Box, Button, Stack, Typography} from "@mui/material";
+            }})
             titleRef.current.innerHTML = note.title;
             contentRef.current.innerHTML = note.content;
             titleRef.current.focus();
```
Я правильно понимаю что этот эффект нужен для инициализации контента? Мне кажется что можно проинициализировать состояние более декларативно, и хук для этого не потребуется.

Нужна будет помощь с тем чтобы эту логику декомпозировать?

nadiia-kobylinska replied:
Да, этот эффект нужен при инициализации существующей редактируемой заметки. Пробовала без эффекта, но как-то не сложилось. 
Подскажи как можно переделать?

Rastopyr replied:
Для редьюсера можно сразу initialState сформировать из `note.content` и `note.title`.
Элементам на которые навешаны `titleRef` и `contentRef` можно передать контент через [`dangerouslySetInnerHTML`](https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml).

А для фокуса есть `autofocus` аттрибут.

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,99 +2,106 @@ import {Box, Button, Stack, Typography} from "@mui/material";
-                        content: content || "empty"
-                    }, hasNoteIndex);
+        dispatch({type: 'change', payload: {loading: true}})
+        try {
```
Какая часть логики тут может выбросить исключение?

nadiia-kobylinska replied:
удалила, так как уже неуместно

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,99 +2,106 @@ import {Box, Button, Stack, Typography} from "@mui/material";
+        dispatch({type: 'change', payload: {loading: true}})
+        try {
+            const hasNoteIndex = props.notes.findIndex((obj => obj.id === state.editMode));
+            props.save({
```
Мне кажется что стоит пропсу назвать `onSave`

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,99 +2,106 @@ import {Box, Button, Stack, Typography} from "@mui/material";
+                    id: hasNoteIndex!==-1 ? state.editMode : Date.now(),
+                    title: state.title || "No name",
+                    content: state.content || "empty"
+                }, hasNoteIndex);
```
А для чего тут используется hasNoteIndex? Мне кажется что там куда это событие отправляется, можно посчитать это значение.

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,99 +2,106 @@ import {Box, Button, Stack, Typography} from "@mui/material";
+            props.save({
+                    id: hasNoteIndex!==-1 ? state.editMode : Date.now(),
+                    title: state.title || "No name",
+                    content: state.content || "empty"
```
Видимо старый код, тк у нас заметки не могут быть пустыми то fallback к дефолтному контенту и title можно убрать, мне кажется

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,99 +2,106 @@ import {Box, Button, Stack, Typography} from "@mui/material";
+        try {
+            const hasNoteIndex = props.notes.findIndex((obj => obj.id === state.editMode));
+            props.save({
+                    id: hasNoteIndex!==-1 ? state.editMode : Date.now(),
```
Глаз зацепился за расчет id.

Тк это дженерал компонент формы, то форма должна просто сабмитить контент, и это стоит отразить в контракте формы. Например `onSubmit`.  А расчет id уже должен быть на более высоком слое, который знает что эта заметка редактируется или сохраняется.

Можно это поправить в рамках другого ПР к след домашке

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,99 +2,106 @@ import {Box, Button, Stack, Typography} from "@mui/material";
+
+    if (props.id && !state.editMode && props.id !== state.editMode){
+        dispatch({type: 'change', payload: {editMode: props.id}})
     }
```
А какое назначение этого диспатча? Может данные, которые он диспатчит, так же можно проинициализировать более декларативно

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,99 +2,106 @@ import {Box, Button, Stack, Typography} from "@mui/material";
+    if (props.id && !state.editMode && props.id !== state.editMode){
+        dispatch({type: 'change', payload: {editMode: props.id}})
     }
+    const isDisabled = state.content === "" || state.title === "" || state.percentCharTitle>100 || state.percentCharContent>100;
```
В редьюсере можно посчитать флаг `isValid`, мне кажется это там должно быть. Это же часть состояния формы.

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -104,18 +111,18 @@ const Form = (props) => {
                     className={"fi=orm-title"}
                     sx={{ mb: 5}}>
-                    {!editMode ? "New Note" : "Edit Note"}
+                    {!state.editMode ? "New Note" : "Edit Note"}
```
nit: Стоит вынести в переменную заголовок

nadiia-emit-digital committed:
Improvement for the validation form

nadiia-kobylinska commented:
@Rastopyr 
Last updates:

- Renamed the variables and event names.
- Simplified logic for better readability. 
- Removed old code.
- Moved ID calculation from the form.
- Updated initialisation and removed unnecessary ‘dispatchs'.
- Cleared duplicate code. Combined contentEditable Divs into one component.
- Added useRef to save cursor position in EditDiv.

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,136 +2,122 @@ import {Box, Button, Stack, Typography} from "@mui/material";
-    let contentRef = useRef(null);
+    const note = props.note;
+
+    const initialState = {
```
Здорово что получилось разгрузить немного тот эффект. Но появилась еще одна проблема, на каждый рендер будет создаваться этот объект. Мелочь а не очень приятно. Мне кажется что это стоит перенести в callback useReducer где определяется начальное состояние

nadiia-kobylinska replied:
Перенесла. 
Столкнулась в результате с проблемой: после сохранения новой заметки не удаляется содержимое с полей, хотя state пустой. А вот если сохранить уже существующую заметку, то всё ок. Предполагаю, что проблема может быть связана с ref и dangerouslySetInnerHTML, но никак не могу найти решение

Rastopyr replied:
Увидел что закомментирован вызов `resetForm`, думаю можно пока вернуть. А в след домашке порефакторить редьюсер и мне кажется что там получится пофиксить проблему

Rastopyr commented on src/components/notes-manage/form-view.js:
```diff
@@ -2,136 +2,122 @@ import {Box, Button, Stack, Typography} from "@mui/material";
-    if (props.id && !editMode && props.id !== editMode){
-        setEditMode(props.id);
+    function onChangeTitle(e){
+        const result = validateTextLength(e,100);
```
В контексте валидации стоит явно передавать валидируемое значение а не событие. Иначе такую фунцию будет немного сложно тестировать.

И что важно, что название функции не согласуется с ее сигнатрой.

nadiia-kobylinska replied:
Заменила на e.target. 
Есть ли смысл здесь использовать ref?

Rastopyr replied:
Не сразу понял почему передается элемент)

Смотри, `calcCountChar` судя по имплементации мутирует элемент(добавляет подстветку). Нарушается контракт функции, то есть она делает не то что обещает, не только валидирует. Так же нарушается принцип единства ответственности.

С точки зрения валидации, функция должна принимать какое то значение, например текст, и возвращать результат валидации. А уже потом на основе результатов этой валидации нужно принимать решение, подсвечивать что то внутри элемента или нет.

Rastopyr replied:
Так же, что меня немного смущает, что эта функция использует DOM api. Мне кажется что стоит использовать react рендеринг

nadiia-kobylinska replied:
подчистила calcCountChar и вынесла подсветку в отдельную функцию (пока отключила её). 
Использовала DOM api так как мне нужно сохранять позицию курсора после подсветки текста.

Rastopyr commented on src/components/notes-manage/formReducer.jsx:
```diff
@@ -0,0 +1,17 @@
+        case 'reset':
+            return action.payload;
+        case 'change':
+            action.payload.isValid = isValid(state);
```
Мутировать данные точно не стоит. Лучше сделать через деструктуризацию строчкой ниже.

nadiia-kobylinska replied:
`return {...state, ...action.payload, isValid: isValid({...state, ...action.payload})};`
я правильно поняла, как-то так?

Rastopyr replied:
```
const nextState = { ...state, ...action.payload }

reutnr { ...nextState, isValid: isValid(nextState)  }
```

Rastopyr replied:
Это если валидировать новый стейт, сейчас валидируется предыдущий

Rastopyr commented on src/components/notes-manage/formReducer.jsx:
```diff
@@ -0,0 +1,17 @@
+    switch (action.type) {
+        case 'reset':
+            return action.payload;
+        case 'change':
```
За этот `change` постояно глаз цепляется. Тут точно должно быть больше событий, но предлагаю в другом ПР)

nadiia-kobylinska replied:
согласна, переделаю)

nadiia-emit-digital committed:
Upd. naming/ moved initialState / fix action

nadiia-emit-digital committed:
upd. calcCountChar
